pkg/backend/metrics.go

```diff
@@ -0,0 +1,222 @@
+package backend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/dns"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+func (ocb *openShiftClusterBackend) emitMetrics(log *logrus.Entry, doc *api.OpenShiftClusterDocument, operationType, provisioningState api.ProvisioningState, backendErr error) error {
+	dimensions := map[string]string{}
+
+	ocb.gatherOperationMetrics(operationType, provisioningState, backendErr, dimensions)
+	ocb.gatherCorrelationID(doc, dimensions)
+	ocb.gatherMiscMetrics(doc, dimensions)
+	ocb.gatherAuthMetrics(doc, dimensions)
+	err := ocb.gatherNetworkMetrics(doc, dimensions)
```

Comment:
We should wrap the error here to make it clear that it is the metrics that is returning, *or* log the error but do not return it and continue along anyway.

Reply:
I've added a warn log here, and a helper function to log warn messages for empty metric values.
This will help to remove duplicated `if` blocks and `log.Warnf()` calls.